rosbag2_transport/src/rosbag2_transport/qos.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "qos.hpp"
+
+namespace YAML
+{
+Node convert<rmw_time_t>::encode(const rmw_time_t & time)
+{
+  Node node;
+  node["sec"] = time.sec;
+  node["nsec"] = time.nsec;
+  return node;
+}
+
+bool convert<rmw_time_t>::decode(const Node & node, rmw_time_t & time)
+{
+  time.sec = node["sec"].as<uint>();
+  time.nsec = node["nsec"].as<uint>();
+  return true;
+}
+
+Node convert<rosbag2_transport::Rosbag2QoS>::encode(const rosbag2_transport::Rosbag2QoS & qos)
+{
+  const auto & p = qos.get_rmw_qos_profile();
+  Node node;
+  node["history"] = static_cast<int>(p.history);
+  node["depth"] = p.depth;
+  node["reliability"] = static_cast<int>(p.reliability);
+  node["durability"] = static_cast<int>(p.durability);
+  node["deadline"] = p.deadline;
+  node["lifespan"] = p.lifespan;
+  node["liveliness"] = static_cast<int>(p.liveliness);
+  node["liveliness_lease_duration"] = p.liveliness_lease_duration;
+  node["avoid_ros_namespace_conventions"] = p.avoid_ros_namespace_conventions;
+  return node;
+}
+
+bool convert<rosbag2_transport::Rosbag2QoS>::decode(
+  const Node & node, rosbag2_transport::Rosbag2QoS & qos)
+{
+  auto history = static_cast<rmw_qos_history_policy_t>(node["history"].as<int>());
+  auto reliability = static_cast<rmw_qos_reliability_policy_t>(node["reliability"].as<int>());
+  auto durability = static_cast<rmw_qos_durability_policy_t>(node["durability"].as<int>());
+  auto liveliness = static_cast<rmw_qos_liveliness_policy_t>(node["liveliness"].as<int>());
+
+  qos
+  .keep_last(node["depth"].as<int>())
+  .history(history)
+  .reliability(reliability)
+  .durability(durability)
+  .deadline(node["deadline"].as<rmw_time_t>())
+  .lifespan(node["lifespan"].as<rmw_time_t>())
+  .liveliness(liveliness)
+  .liveliness_lease_duration(node["liveliness_lease_duration"].as<rmw_time_t>())
+  .avoid_ros_namespace_conventions(node["avoid_ros_namespace_conventions"].as<bool>());
```

Comment:
I know it's a hard problem to solve, but any idea about how we could try to keep this in sync as qos gets new fields?

Reply:
I don't have a good answer to this. My thought as of right now is just that rosbag2 needs to have a pass to support new fields if they are added, and until that time it will only support the default values for any given policy.

I'm not aware of a way to introspect fields of a struct in C++, but that doesn't mean there isn't a way

Reply:
Ah - I thought of something we can do. It is a "change detector" like we try to avoid, but in this case of serialization method, it seems like the right thing to do. I'm adding a test that explicitly builds a rmw_qos_profile_t using {} initializer, so if new fields are added, the build will fail.

Reply:
Thinking about it, `deserialize(serialize(x) == x` should do the trick?

Reply:
If you're sure the qos struct will always be a POD, memcmp is an option otherwise (tread carefully as it can blow up easily)

Reply:
I don't think it will. If new fields are added - when I construct the object it will use the default values for those new fields. Now when I serialize, it will write out all the fields we specified here. Then, on deserializing, it will read the fields that it knows, with the unknown fields having default values. The equality will pass and new fields will have slipped through undetected

Reply:
we could maybe apply something like a [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) which only allows to build the object if all parts are present. But I don't see a real way around updating the code base if the QoS struct gets more fields.